# Домашнее задание к лекции Улучшение качества модели

Взять boston house-prices datase (sklearn.datasets.load_boston), для задачи регрессии попробовать разные алгоритмы, поподбирать параметры, вывести итоговое качество

In [37]:
# Подключим датасет и библиотеки
from sklearn.datasets import load_boston
import numpy as np
import warnings

warnings.filterwarnings('ignore')

In [38]:
# Загрузим данные
with warnings.catch_warnings():
    warnings.filterwarnings("ignore")
    X, y = load_boston(return_X_y=True)

In [39]:
# Подключим библиотеки регрессии
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import ExtraTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neighbors import RadiusNeighborsRegressor

In [40]:
lr_model = LinearRegression()
dt_model = DecisionTreeRegressor()
et_model = ExtraTreeRegressor()
kn_model = KNeighborsRegressor()
rn_model = RadiusNeighborsRegressor()

In [48]:
lr_params = {'fit_intercept':[True, False], 'normalize':[True, False]}
dt_params = {'min_samples_split':[x for x in range(1,10)], 'min_samples_leaf':[x for x in range(1,10)]}
et_params = {'min_samples_split':[x for x in range(1,10)], 'min_samples_leaf':[x for x in range(1,10)]}
kn_params = {'n_neighbors': [x for x in range(1,10)], 'weights': ['uniform', 'distance'], 'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute']}
rn_params = {'radius': [x for x in np.arange(0.0, 5.1, 0.2)], 'weights': ['uniform', 'distance'], 'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute']}


In [49]:
# Сделаем списки с моделями и параметрами для них
list_model = [lr_model, dt_model, et_model, kn_model, rn_model]
list_param = [lr_params, dt_params, et_params, kn_params, rn_params]

In [65]:
# Подбирать параметры будем через GridSearchCV - она делает все что нужно
from sklearn.model_selection import GridSearchCV

for model, grid_param in zip(list_model, list_param):
    grid = GridSearchCV(model, grid_param, cv=10, scoring='r2')
    grid.fit(X, y)
    print(model)
    print("Best score ->", grid.best_score_)
    print("Best params ->", grid.best_params_)
    print("Best estimator ->", grid.best_estimator_)
    for k in grid.cv_results_:
        print(k, ":", grid.cv_results_[k][0])
    
    print("-----------------------------------------------------")

LinearRegression()
Best score -> 0.2760345915425497
Best params -> {'fit_intercept': False, 'normalize': True}
Best estimator -> LinearRegression(fit_intercept=False, normalize=True)
mean_fit_time : 0.0017441034317016602
std_fit_time : 0.0004957044389366591
mean_score_time : 0.0006300926208496094
std_score_time : 0.000145588458075346
param_fit_intercept : True
param_normalize : True
params : {'fit_intercept': True, 'normalize': True}
split0_test_score : 0.733760819350227
split1_test_score : 0.4730725033143366
split2_test_score : -1.006314543726404
split3_test_score : 0.6411398421934933
split4_test_score : 0.5476604617627834
split5_test_score : 0.7364029190694071
split6_test_score : 0.3782838635247935
split7_test_score : -0.12922703208877095
split8_test_score : -0.768432431438508
split9_test_score : 0.41894349864428415
mean_test_score : 0.20252899006056416
std_test_score : 0.5952960169512276
rank_test_score : 3
-----------------------------------------------------
DecisionTreeRegressor(

### Итоги:
Лучше всего справилась ExtraTreeRegressor, при этом best params min_samples_leaf: 4, min_samples_split: 3. Попробуем построить с такими параметрами модель ExtraTreeRegressor

In [104]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

In [105]:
et_model_2 = ExtraTreeRegressor(min_samples_leaf=4, min_samples_split=3)

In [106]:
et_model_2.fit(X_train, y_train)

ExtraTreeRegressor(min_samples_leaf=4, min_samples_split=3)

In [107]:
et_model_2.score(X_test, y_test)

0.7050101226051871

Качество модели 0.705

Построим для интереса линейную регрессию с базовыми параметрами

In [108]:
lr_model_2 = LinearRegression()
lr_model_2.fit(X_train, y_train)

LinearRegression()

In [109]:
lr_model_2.score(X_test, y_test)

0.7789207451814436

Качество даже выше